metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py

```diff
@@ -0,0 +1,911 @@
+                    message = f"API({scan_get_endpoint}) return error code {res.status_code} for scan id({scan_id})"
+
+                    LOGGER.warning(message)
+
+                    raise ConnectionError(message)
+
+                if res.json()["status"].upper() == "Succeeded".upper():
+                    LOGGER.info(f"Scan result is available for scan id({scan_id})")
+                    return True
+
+                if trail == max_trial:
+                    break
+                LOGGER.info(f"Sleeping for {minimum_sleep} seconds")
+                sleep(minimum_sleep)
+                trail += 1
+
+            # Result is not available
+            return False
+
+        def get_scan_result(scan_id: str) -> dict:
+            LOGGER.info("Fetching scan  result")
+            LOGGER.info(f"{Constant.SCAN_ID}={scan_id}")
+            scan_result_get_endpoint = PowerBiAPI.API_ENDPOINTS[
+                Constant.SCAN_RESULT_GET
+            ]
+            scan_result_get_endpoint = scan_result_get_endpoint.format(
+                POWERBI_ADMIN_BASE_URL=PowerBiAPI.ADMIN_BASE_URL, SCAN_ID=scan_id
+            )
+
+            LOGGER.info(f"Hitting URL={scan_result_get_endpoint}")
+            res = requests.get(
+                scan_result_get_endpoint,
+                headers={Constant.Authorization: self.get_access_token()},
+            )
+            if res.status_code != 200:
+                message = f"API({scan_result_get_endpoint}) return error code {res.status_code} for scan id({scan_id})"
+
+                LOGGER.warning(message)
+
+                raise ConnectionError(message)
+
+            return res.json()["workspaces"][0]
+
+        def json_to_dataset_map(scan_result: dict) -> dict:
+            """
+            Filter out "dataset" from scan_result and return PowerBiAPI.Dataset instance set
+            """
+            datasets: Optional[Any] = scan_result.get("datasets")
+            dataset_map: dict = {}
+
+            if datasets is None or len(datasets) == 0:
+                LOGGER.warning(
+                    f'Workspace {scan_result["name"]}({scan_result["id"]}) does not have datasets'
+                )
+
+                LOGGER.info("Returning empty datasets")
+                return dataset_map
+
+            for dataset_dict in datasets:
+                dataset_instance: PowerBiAPI.PowerBIDataset = self.get_dataset(
+                    workspace_id=scan_result["id"],
+                    dataset_id=dataset_dict["id"],
+                )
+                dataset_map[dataset_instance.id] = dataset_instance
+                # set dataset-name
+                dataset_name: str = (
+                    dataset_instance.name
+                    if dataset_instance.name is not None
+                    else dataset_instance.id
+                )
+
+                for table in dataset_dict["tables"]:
+                    expression: str = (
+                        table["source"][0]["expression"]
```

Comment:
is the first element in the list guaranteed to be present?

Reply:
As per document it is guaranteed to be present 
https://learn.microsoft.com/en-us/rest/api/power-bi/admin/workspace-info-get-scan-result#table

Reply:
we should not trust them :)